File: biased_decisions/compliance_test.py
"""Specs for the compliance block of the data contract (``biased_decisions.compliance``): the
regulated-practice mapping, the citations it rests on, and the evidence every warning, panel,
failure recipe and guidance insight on the site points to. Each spec names the failure it
prevents: a claim with no measured cell behind it, a citation the repository never verified, or
a number that drifted from the record."""
from __future__ import annotations

import json
import re

import pytest

from biased_decisions.compliance import _one_laya, normalise
from biased_decisions.leaderboard import generate_json
from biased_decisions.tasks.base import DEFAULT_ROOT


@pytest.fixture(scope="module")
def doc():
    return generate_json(DEFAULT_ROOT, date="2026-09-23")


@pytest.fixture(scope="module")
def comp(doc):
    return doc["compliance"]


def _facet(doc, ev):
    dim = next(d for d in doc["dimensions"] if d["id"] == ev["dimension"])
    cell = next(c for c in dim["breakdown"]["cells"]
                if c["group"] == ev["group"] and c["item"] == ev["item"])
    return dim, cell, cell["engines"][ev["engine"]]


def test_the_contract_carries_a_compliance_block_under_a_new_schema(doc):
    assert doc["schema"] == "biased-decisions/leaderboard@4"
    for key in ("notice", "citations", "practices", "mapping", "evidence", "shortlist",
                "recipes", "insights", "checklist", "articles", "unmeasured", "inversion"):
        assert key in doc["compliance"], key


def test_the_notice_says_plainly_that_this_is_not_legal_advice(comp):
    assert "not legal advice" in comp["notice"].lower()


def test_every_citation_is_one_the_repository_already_names_verbatim(comp):
    for c in comp["citations"]:
        ref = c["repo_ref"]
        source = normalise((DEFAULT_ROOT / ref["path"]).read_text(encoding="utf-8"))
        assert normalise(ref["text"]) in source, f"{c['id']}: {ref['text']!r} not in {ref['path']}"
        assert c["primary_url"].startswith("https://"), c["id"]
        assert c["quotes"], f"{c['id']}: no quoted primary text"
        for q in c["quotes"]:
            assert q["text"] and q["where"] and q["source_url"].startswith("https://")


def test_a_regime_the_repository_only_names_is_listed_as_unmeasured_never_warned_on(comp):
    cited = {c["id"] for c in comp["citations"]}
    for u in comp["unmeasured"]:
        source = normalise((DEFAULT_ROOT / u["repo_ref"]["path"]).read_text(encoding="utf-8"))
        assert normalise(u["repo_ref"]["text"]) in source, u["id"]
        assert u["id"] not in cited
    for m in comp["mapping"]:
        assert not set(m.get("citations", [])) & {u["id"] for u in comp["unmeasured"]}


def test_every_dimension_is_mapped_once_and_every_regulated_one_cites_a_verified_rule(doc, comp):
    ids = [m["dimension"] for m in comp["mapping"]]
    assert sorted(ids) == sorted(d["id"] for d in doc["dimensions"])
    cited = {c["id"] for c in comp["citations"]}
    practices = {p["id"] for p in comp["practices"]}
    recipes = {r["id"] for r in comp["recipes"]}
    insights = {i["id"] for i in comp["insights"]}
    for m in comp["mapping"]:
        if not m["regulated"]:
            assert m["reason"], m["dimension"]
            continue
        assert m["practice"] in practices
        assert m["citations"] and set(m["citations"]) <= cited, m["dimension"]
        for key in ("decision", "attribute", "failure_mode", "who_harmed"):
            assert m[key], (m["dimension"], key)
        assert m["recipes"] and set(m["recipes"]) <= recipes, m["dimension"]
        assert m["insights"] and set(m["insights"]) <= insights, m["dimension"]


def test_option_order_is_not_flagged_as_a_regulated_decision(comp):
    m = next(m for m in comp["mapping"] if m["dimension"] == "option-order")
    assert m["regulated"] is False


def test_every_cell_evidence_entry_copies_a_measured_facet_number_for_number(doc, comp):
    for ev in comp["evidence"]:
        if ev["kind"] != "cell":
            continue
        dim, cell, f = _facet(doc, ev)
        assert f["status"] == "measured", ev["id"]
        assert (ev["value"], ev["lo"], ev["hi"]) == (f["raw"]["value"], f["raw"]["lo"],
                                                     f["raw"]["hi"]), ev["id"]
        assert ev["n"] == f["n"] and ev["floor"] == f["floor"]["value"], ev["id"]
        assert ev["excess"] == f["excess"] and ev["detected"] == f["detected"], ev["id"]
        assert ev["has_example"] == bool(cell["example"]), ev["id"]


def test_the_shortlist_block_is_the_replayed_record_row_for_row(comp):
    rows = []
    for block in comp["shortlist"]["pairs"]:
        path = DEFAULT_ROOT / block["study"]
        want = _one_laya([json.loads(line) for line in path.read_text(encoding="utf-8").splitlines() if line])
        assert block["rows"] == want, block["task"]
        rows.extend(block["rows"])
    assert rows
    assert comp["shortlist"]["line"] == 0.8


def test_every_shortlist_and_prereg_evidence_entry_resolves(comp):
    rows = {(b["task"], r["engine"], r["variant"], r["cut"])
            for b in comp["shortlist"]["pairs"] for r in b["rows"]}
    prereg = (DEFAULT_ROOT / "studies/PREREGISTERED.md").read_text(encoding="utf-8")
    for ev in comp["evidence"]:
        if ev["kind"] == "shortlist":
            assert (ev["task"], ev["engine"], ev["variant"], ev["cut"]) in rows, ev["id"]
        elif ev["kind"] == "prereg":
            for key in ("measurement", "prediction", "observed", "verdict"):
                assert ev[key], (ev["id"], key)
            assert ev["measurement"] in prereg.replace("**", "")
        elif ev["kind"] == "floor":
            assert ev["value"] is not None and ev["n"]
        else:
            assert ev["kind"] == "cell", ev["id"]


def test_every_recipe_and_insight_stands_on_evidence_and_links_both_ways(comp):
    evidence = {e["id"] for e in comp["evidence"]}
    insights = {i["id"] for i in comp["insights"]}
    practices = {p["id"] for p in comp["practices"]}
    slug = re.compile(r"^[a-z0-9]+(?:-[a-z0-9]+)*$")
    seen = set()
    for r in comp["recipes"]:
        assert slug.match(r["id"]) and r["id"] not in seen, r["id"]
        seen.add(r["id"])
        assert r["evidence"] and set(r["evidence"]) <= evidence, r["id"]
        assert r["practices"] and set(r["practices"]) <= practices | {"any"}, r["id"]
        assert r["pattern"] and r["what_happens"] and r["inverse"], r["id"]
        assert r["insight"] in insights, r["id"]
    for i in comp["insights"]:
        assert slug.match(i["id"]) and i["id"] not in seen, i["id"]
        seen.add(i["id"])
        assert (set(i["evidence"]) <= evidence) and (i["evidence"] or i["links"]), i["id"]
    for c in comp["checklist"]:
        assert c["insight"] in insights, c["text"]


def test_prose_carries_no_numbers_so_every_number_on_a_compliance_page_comes_from_evidence(comp):
    digit = re.compile(r"\d")
    for r in comp["recipes"]:
        for key in ("title", "pattern", "what_happens", "inverse"):
            assert not digit.search(r[key]), (r["id"], key, r[key])
    for i in comp["insights"]:
        assert not digit.search(i["title"]) and not digit.search(i["text"]), i["id"]
    for m in comp["mapping"]:
        for key in ("failure_mode", "who_harmed"):
            if m.get(key):
                assert not digit.search(m[key].replace("4(a)", "").replace("4(b)", "")), (m["dimension"], key)


def test_risks_are_phrased_as_exposure_never_as_findings_of_illegality(comp):
    banned = re.compile(r"\b(is|are|was|were) (illegal|unlawful)\b|\bviolat(es|ed|ion)\b|\bbreaks the law\b", re.I)
    prose = [comp["notice"], comp["inversion"]["lede"]]
    prose += [r[k] for r in comp["recipes"] for k in ("title", "pattern", "what_happens", "inverse")]
    prose += [i["text"] for i in comp["insights"]]
    prose += [m.get(k) or "" for m in comp["mapping"] for k in ("failure_mode", "who_harmed", "reason")]
    for text in prose:
        assert not banned.search(text), text


def test_the_inversion_attribution_is_sourced(comp):
    inv = comp["inversion"]
    assert inv["quote"] == "Invert, always invert."
    assert "Jacobi" in inv["attribution"] and "Munger" in inv["attribution"]
    assert inv["source_url"].startswith("https://")


def test_the_companion_articles_are_linked(comp):
    urls = {a["url"] for a in comp["articles"]}
    assert {"https://anth.us/blog/encoding-prejudice/", "https://anth.us/blog/one-word-test/",
            "https://anth.us/blog/can-you-fix-it/"} <= urls

File: site/test/compliance.test.mjs
// Specs for the compliance pages and panels (BD-7ee753), run after `astro build`. They read
// dist/ and the data contract only. Each names the failure it prevents: a warning with no
// measured result behind it, a citation with no link, a deep link that lands nowhere, or a
// compliance page that does not say it is not legal advice.
import { test } from "node:test";
import assert from "node:assert/strict";
import { readFileSync, readdirSync, statSync, existsSync } from "node:fs";
import { join, relative, sep } from "node:path";
import { fileURLToPath } from "node:url";

// Decisions that are not hiring: pages for these carry no hiring warning (see NON_HIRING_ITEMS in site.js).
const NON_HIRING = ["qpain-treatment", "civil-comments-moderation", "tenant-inquiry-viewing", "small-business-loan",
  "cfpb-escalate-servicemember", "cfpb-escalate-older", "cfpb-escalate-family"];

const DIST = fileURLToPath(new URL("../dist/", import.meta.url));
const DATA = JSON.parse(readFileSync(new URL("../data/leaderboard.json", import.meta.url), "utf8"));
const C = DATA.compliance;

function walk(dir, out = []) {
  for (const name of readdirSync(dir)) {
    const p = join(dir, name);
    if (statSync(p).isDirectory()) walk(p, out); else out.push(p);
  }
  return out;
}
const urlOf = (file) => "/" + relative(DIST, file).split(sep).join("/").replace(/index\.html$/, "");
const pages = new Map(walk(DIST).filter((f) => f.endsWith("index.html")).map((f) => [urlOf(f), readFileSync(f, "utf8")]).filter(([, h]) => !h.includes('http-equiv="refresh"')));
const unescape = (s) => s.replace(/&quot;/g, '"').replace(/&#39;/g, "'").replace(/&lt;/g, "<").replace(/&gt;/g, ">").replace(/&amp;/g, "&");
const text = (html) => unescape(html.replace(/<script[\s\S]*?<\/script>/g, " ").replace(/<[^>]+>/g, " ")).replace(/\s+/g, " ");
const ids = (html) => new Set([...html.matchAll(/\sid="([^"]+)"/g)].map((m) => m[1]));

// The HTML of the element that opens at `start` (a tag with the given name), up to its matching close.
function element(html, start, tag) {
  const re = new RegExp(`<${tag}\\b|</${tag}>`, "g");
  re.lastIndex = start;
  let depth = 0, m;
  while ((m = re.exec(html))) {
    depth += m[0].startsWith("</") ? -1 : 1;
    if (depth === 0) return html.slice(start, m.index + m[0].length);
  }
  return html.slice(start);
}
function byId(html, id) {
  const m = new RegExp(`<(\\w+)[^>]*\\sid="${id}"`).exec(html);
  return m ? element(html, m.index, m[1]) : null;
}

// Where a link lands: the page must exist, and a fragment must name an element on it.
function landing(href, from) {
  const [path, frag] = href.split("#");
  const target = path === "" ? from : path;
  const html = pages.get(target);
  if (!html) return `${from} -> ${href}: no page`;
  if (frag && !ids(html).has(frag)) return `${from} -> ${href}: no #${frag} there`;
  return null;
}

const mapping = Object.fromEntries(C.mapping.map((m) => [m.dimension, m]));
const regulated = C.mapping.filter((m) => m.regulated).map((m) => m.dimension);
const dimOfPath = (p) => {
  const segs = p.split("/").filter(Boolean);
  if (segs[0] === "engines") return segs.length === 3 ? segs[2] : null;
  if (segs.slice(1).some((x) => NON_HIRING.includes(x))) return null;
  return DATA.dimensions.some((d) => d.id === segs[0]) ? segs[0] : null;
};
const primary = new Set(C.citations.map((c) => c.primary_url));

test("every page of a regulated dimension shows a red warning that is not colour alone", () => {
  let n = 0;
  for (const [path, html] of pages) {
    const dim = dimOfPath(path);
    if (!dim || !regulated.includes(dim)) continue;
    n++;
    const m = /<a class="reg-warn[^"]*" href="#risk" aria-label="([^"]+)"[^>]*>([\s\S]*?)<\/a>/.exec(html);
    assert.ok(m, `${path}: no warning linking to its risk panel`);
    assert.match(m[1], /^Regulated decision: /, `${path}: warning has no accessible name`);
    assert.match(m[2], /<svg[^>]*aria-hidden="true"/, `${path}: warning has no icon`);
    assert.match(text(m[2]), /Regulated decision/, `${path}: warning has no words`);
  }
  assert.ok(n > 100, `only ${n} regulated pages found`);
});

test("pages of an unregulated dimension carry no warning", () => {
  for (const [path, html] of pages) {
    const dim = dimOfPath(path);
    if (dim && !mapping[dim].regulated) assert.doesNotMatch(html, /class="reg-warn/, path);
  }
});

test("every risk panel links a measured result, a citation, a failure recipe and a guidance insight", () => {
  const recipes = new Set(C.recipes.map((r) => r.id));
  const insights = new Set(C.insights.map((i) => i.id));
  for (const [path, html] of pages) {
    const dim = dimOfPath(path);
    if (!dim || !regulated.includes(dim)) continue;
    const panel = byId(html, "risk");
    assert.ok(panel, `${path}: no risk panel`);
    assert.match(text(panel), /not legal advice/i, `${path}: panel does not say it is not legal advice`);
    const ev = [...panel.matchAll(/<a class="ev-link" href="([^"]+)"/g)].map((m) => m[1]);
    assert.ok(ev.length, `${path}: panel states no measured result`);
    for (const href of ev) assert.equal(landing(href, path), null);
    assert.match(text(panel), /between [−-]?\d+\.\d+ and [−-]?\d+\.\d+/, `${path}: panel quotes no interval`);
    assert.match(text(panel), /\d[\d,]* (?:bios|texts|comments|case descriptions|complaints|rental inquiries|loan applications|resume summaries|women)\b/, `${path}: panel quotes no sample size`);
    const cites = [...panel.matchAll(/<a class="cite" href="([^"]+)"/g)].map((m) => m[1]);
    assert.ok(cites.length, `${path}: panel cites no rule`);
    for (const c of cites) assert.ok(primary.has(c), `${path}: ${c} is not a verified citation`);
    const rec = [...panel.matchAll(/href="\/how-to-fail\/#([^"]+)"/g)].map((m) => m[1]);
    const ins = [...panel.matchAll(/href="\/guidance\/#([^"]+)"/g)].map((m) => m[1]);
    assert.ok(rec.length && ins.length, `${path}: panel links no recipe or no insight`);
    for (const r of rec) assert.ok(recipes.has(r), `${path}: unknown recipe ${r}`);
    for (const i of ins) assert.ok(insights.has(i), `${path}: unknown insight ${i}`);
  }
});

test("the inversion page grounds every recipe and gives each a stable anchor", () => {
  const html = pages.get("/how-to-fail/");
  assert.ok(html, "no /how-to-fail/ page");
  const t = text(html);
  assert.match(t, /not legal advice/i);
  assert.ok(t.includes(C.inversion.quote), "no Munger quote");
  assert.ok(html.includes(`href="${C.inversion.source_url}"`), "the quote is not sourced");
  for (const r of C.recipes) {
    const art = byId(html, r.id);
    assert.ok(art, `no #${r.id}`);
    const ev = [...art.matchAll(/<a class="ev-link" href="([^"]+)"/g)].map((m) => m[1]);
    assert.ok(ev.length >= r.evidence.length, `#${r.id}: ${ev.length} evidence links for ${r.evidence.length} entries`);
    for (const href of ev) {
      if (href.startsWith("http")) continue;
      assert.equal(landing(href, "/how-to-fail/"), null);
    }
    assert.ok(text(art).includes(r.inverse), `#${r.id}: no inverse practice`);
    assert.match(art, new RegExp(`href="/guidance/#${r.insight}"`), `#${r.id}: no guidance link`);
  }
});

test("the guidance page links every insight to its evidence and carries the checklist", () => {
  const html = pages.get("/guidance/");
  assert.ok(html, "no /guidance/ page");
  assert.match(text(html), /not legal advice/i);
  for (const i of C.insights) {
    const sec = byId(html, i.id);
    assert.ok(sec, `no #${i.id}`);
    const links = [...sec.matchAll(/<a class="ev-link" href="([^"]+)"/g)].map((m) => m[1]);
    assert.ok(links.length >= i.evidence.length, `#${i.id}: evidence not linked`);
    for (const href of links) if (!href.startsWith("http")) assert.equal(landing(href, "/guidance/"), null);
  }
  const list = byId(html, "checklist");
  assert.ok(list, "no checklist");
  for (const c of C.checklist) assert.ok(text(list).includes(c.text), c.text);
  for (const a of C.articles) assert.ok(html.includes(`href="${a.url}"`), a.url);
});

test("every evidence entry quoted on a page shows the record's number", () => {
  const fmt = (v) => Number(v).toFixed(2);
  const byEv = Object.fromEntries(C.evidence.map((e) => [e.id, e]));
  let n = 0;
  for (const [path, html] of pages) {
    for (const m of html.matchAll(/<li class="ev[^"]*" data-ev="([^"]+)"/g)) {
      const e = byEv[m[1]];
      assert.ok(e, `${path}: unknown evidence ${m[1]}`);
      const li = text(element(html, m.index, "li")).replace(/−/g, "-");
      if (e.kind === "cell") assert.ok(li.includes(`between ${fmt(e.lo)} and ${fmt(e.hi)}`) && li.includes(fmt(e.value)), `${path}: ${e.id} shows ${li}`);
      if (e.kind === "prereg") assert.ok(li.includes(e.observed.replace(/\*/g, "").slice(0, 20)), `${path}: ${e.id}`);
      n++;
    }
  }
  assert.ok(n > 0);
});

test("the shortlist is on its task pages, every row deep-linkable", () => {
  for (const block of C.shortlist.pairs) {
    const path = `/gender/${block.task}/`;
    const html = pages.get(path);
    const sec = byId(html, "shortlist");
    assert.ok(sec, `${path}: no #shortlist`);
    for (const r of block.rows) assert.ok(ids(html).has(`shortlist-${r.engine}-${r.variant.replace(/_/g, "-")}-${r.cut}`), `${path}: row ${r.engine} ${r.variant} ${r.cut}`);
    assert.match(text(sec), /four-fifths/);
  }
});

test("the home page points to the inversion and guidance pages and flags regulated dimensions", () => {
  const html = pages.get("/");
  assert.match(html, /href="\/how-to-fail\/"/);
  assert.match(html, /href="\/guidance\/"/);
  const flags = [...html.matchAll(/class="reg-warn compact"/g)].length;
  assert.ok(flags >= regulated.length, `${flags} compact flags for ${regulated.length} regulated dimensions`);
});

test("the masthead links the compliance pages from every page", () => {
  for (const [path, html] of pages) {
    if (path === "/og-gallery/" || !html.includes('class="topnav"')) continue;
    assert.match(html, /<nav class="topnav"[\s\S]*href="\/how-to-fail\/"[\s\S]*<\/nav>/, path);
  }
});

test("a page for the opioid or comment tasks carries no hiring warning", () => {
  let n = 0;
  for (const [path, html] of pages) {
    const segs = path.split("/").filter(Boolean);
    if (segs[0] === "engines" || !segs.slice(1).some((x) => NON_HIRING.includes(x))) continue;
    n++;
    assert.doesNotMatch(html, /class="reg-warn/, path);
  }
  assert.ok(n > 5, `only ${n} such pages found`);
});
